lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java

```diff
@@ -0,0 +1,155 @@
+  }
+
+
+  @Override
+  public void channelActive(final ChannelHandlerContext ctx) {
+    final Channel channel = ctx.channel();
+
+    if (LOG.isLoggable(Level.FINEST)) {
+      LOG.log(Level.FINEST, "Channel active. key: {0}", channel.remoteAddress());
+    }
+
+    this.addrToLinkRefMap.putIfAbsent(
+        channel.remoteAddress(), new LinkReference(new NettyLink<>(
+            channel, new ByteCodec(), new LoggingLinkListener<byte[]>(), uri)));
+
+    LOG.log(Level.FINER, "Add connected channel ref: {0}", this.addrToLinkRefMap.get(channel.remoteAddress()));
+
+  }
+
+  @Override
+  public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
+    if(msg instanceof HttpRequest) {
+      LOG.log(Level.FINEST, "HttpRequest received");
+
+      final HttpRequest request = (HttpRequest) msg;
+      final HttpHeaders headers = request.headers();
+      this.httpRequest = request;
+
+      if (!headers.isEmpty()) {
+        for (final Map.Entry<String, String> h: headers) {
+          final CharSequence key = h.getKey();
+          final CharSequence value = h.getValue();
+          buf.append("HEADER: ").append(key).append(" = ").append(value).append("\r\n");
+        }
+        buf.append("\r\n");
+      }
+      appendDecoderResult(buf, request);
+    } else if (msg instanceof HttpContent) {
+      LOG.log(Level.FINEST, "HttpContent received");
+      final HttpContent httpContent = (HttpContent) msg;
+      final ByteBuf content = httpContent.content();
+      if (content.isReadable()) {
+        buf.append("CONTENT: ");
+        buf.append(content.toString(CharsetUtil.UTF_8));
+        buf.append("\r\n");
+        appendDecoderResult(buf, this.httpRequest);
+      }
+
+      if (msg instanceof LastHttpContent) {
+        buf.append("END OF CONTENT\r\n");
+        final LastHttpContent trailer = (LastHttpContent) msg;
+        if (!trailer.trailingHeaders().isEmpty()) {
+          buf.append("\r\n");
+          for (CharSequence name: trailer.trailingHeaders().names()) {
+            for (CharSequence value: trailer.trailingHeaders().getAll(name)) {
+              buf.append("TRAILING HEADER: ");
+              buf.append(name).append(" = ").append(value).append("\r\n");
+            }
+          }
+          buf.append("\r\n");
+        }
+
+        final Channel channel = ctx.channel();
+        final byte[] message = new byte[content.readableBytes()];
+        content.readBytes(message);
+        if (LOG.isLoggable(Level.FINEST)) {
+          LOG.log(Level.FINEST, "MessageEvent: local: {0} remote: {1} :: {2}", new Object[]{
+                  channel.localAddress(), channel.remoteAddress(), content});
+        }
+
+        if (message.length > 0) {
+          // send to the dispatch stage
+          this.stage.onNext(this.getTransportEvent(message, channel));
+        }
+      }
+    } else {
+      LOG.log(Level.SEVERE, "Unknown type of message received: {0}", msg);
+    }
+    LOG.log(Level.FINEST, buf.toString());
```

Comment:
Also, same as above: if we use `buf` for logging purposes only, make sure you append to it only when the logging level is low enough.